### PR TITLE
 rbd: add generic ephemeral volume validation

### PR DESCRIPTION
**Pull request by humblec**
rbd: add generic ephemeral volume validation
    
    This commit adds the validation of csi RBD driver to work with
    ephemeral volume support. With ephemeral volume support a user can
    specify ephemeral volumes in its pod spec and tie the lifecycle
    of the PVC with the POD.
    
    An example POD spec looks like this:
    
    kind: Pod
    apiVersion: v1
    metadata:
      name: my-app
       spec:
      containers:
        - name: csi-rbd-demo-pod
          image: busybox
          volumeMounts:
            - mountPath: "/scratch"
              name: mypvc
          command: [ "sleep", "1000000" ]
      volumes:
        - name: mypvc
          ephemeral:
            volumeClaimTemplate:
              spec:
                accessModes: [ "ReadWriteOnce" ]
                storageClassName: "csi-rbd-sc"
                resources:
                  requests:
                    storage: 1Gi

Additional Ref # https://github.com/rook/rook/pull/9055

Updates #2587 

Signed-off-by: Humble Chirammal <hchiramm@redhat.com>

### CODE REVIEW

File: e2e/rbd.go [hunk 1]
```diff
@@ -367,6 +368,28 @@ var _ = Describe("RBD", func() {
 					}
 				})
 			}
+			By("verify ephemeral volume support", func() {
+				// create application
+				app, err := loadApp(appEphemeralPath)
+				if err != nil {
+					e2elog.Failf("failed to load application with error %v", err)
```
Comment (Madhu-1) [inline e2e/rbd.go]: remove with error from all places

File: e2e/rbd.go [hunk 2]
```diff
@@ -367,6 +368,28 @@ var _ = Describe("RBD", func() {
...
+				}
+				// validate created backend rbd images
+				validateRBDImageCount(f, 1, defaultRBDPool)
+
+				err = deletePod(app.Name, app.Namespace, f.ClientSet, deployTimeout)
+				if err != nil {
+					e2elog.Failf("failed to delete application with error %v", err)
+				}
+				// validate created backend rbd images
+				validateRBDImageCount(f, 0, defaultRBDPool)
```
Comment (Madhu-1) [inline e2e/rbd.go:393]: please validate image in trash

File: examples/rbd/pod-ephemeral.yaml [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+      volumeMounts:
+        - mountPath: "/myspace"
+          name: mypvc
+  volumes:
+    - name: mypvc
+      ephemeral:
+        volumeClaimTemplate:
+          spec:
+            accessModes: ["ReadWriteOnce"]
+            storageClassName: "csi-rbd-sc"
```
Comment (Madhu-1) [inline examples/rbd/pod-ephemeral.yaml]: quotes not required for mountPath and storageclass and also for access mode? please keep the same as in other yamls.

File: e2e/rbd.go [hunk 4]
```diff
@@ -367,6 +368,28 @@ var _ = Describe("RBD", func() {
 					}
 				})
 			}
+			By("verify ephemeral volume support", func() {
```
Comment (Madhu-1) [inline e2e/rbd.go]: generic ephemeral volume support?

File: e2e/rbd.go [hunk 5]
```diff
@@ -367,6 +368,28 @@ var _ = Describe("RBD", func() {
...
+				// create application
+				app, err := loadApp(appEphemeralPath)
```
Comment (Madhu-1) [inline e2e/rbd.go:377]: CI will fail for 20 as the feature gate is still alpha?

* add kube validation check here
* enable feature gate for kube 1.20?

**Review by Madhu-1 — DISMISSED** [review]

**Comment by Madhu-1** [comment]
please add same e2e for cephfs also

File: e2e/rbd.go (hunk 5 above)
Comment (humblec) [reply e2e/rbd.go:377]: running the tests only >= 21 is good enough I believe. wdyt?

Comment (Madhu-1) [reply e2e/rbd.go:377]: my suggestion is to run from 20+ as we test on kube 1.20,21, and 22 also

**Comment by humblec** [comment]
@Madhu-1 added tests for cephfs too and enabled feature gate only for kube versions >=1.20.  PTAL.

File: scripts/minikube.sh [hunk 6]
```diff
@@ -202,6 +211,14 @@ up)
 
     disable_storage_addons
 
+    #  get kubernetes version we are operating on and accordingly enable feature gates
+    KUBE_MAJOR=$(kube_version 1)
+    KUBE_MINOR=$(kube_version 2)
+    if  [ "${KUBE_MAJOR}" -eq 1 ] && [ "${KUBE_MINOR}" -ge 20 ];
```
Comment (nixpanic) [inline scripts/minikube.sh]: 1. not sure you can compare strings with `-eq` or `-ge`
2. no `;` needed when `then` is on the next  line

File: scripts/minikube.sh [hunk 7]
```diff
@@ -202,6 +211,14 @@ up)
...
+    then
+    # if kubernetes version is greater than 1.20 enable Generic Ephemeral Volume feature gate
+    K8S_FEATURE_GATES="GenericEphemeralVolume=true"
```
Comment (nixpanic) [inline scripts/minikube.sh]: `K8S_FEATURE_GATES` is already set at line 163, this overrides that option. Should the variable not be extended with the feature gate instead?

**Review by nixpanic — CHANGES_REQUESTED** [review]
It is simpler to included the example .yaml files with the e2e commits so that they are self contained and really belong together. There is then no need to include the full yaml in the commit message anymore, cleaning things up a little.

**Comment by humblec** [comment]
> It is simpler to included the example .yaml files with the e2e commits so that they are self contained and really belong together. There is then no need to include the full yaml in the commit message anymore, cleaning things up a little.

sure, squashed commits of example yaml to the validation commit.

File: scripts/minikube.sh (hunk 6 above)
Comment (humblec) [reply scripts/minikube.sh]: the `-eq or ge` is working as expected, I have removed unwanted `;` in the last refresh.

File: scripts/minikube.sh (hunk 7 above)
Comment (humblec) [reply scripts/minikube.sh]: my plan was to remove the expandCSIVolumes feature gate as a follow up as it no longer wanted, will file a seperate PR for the same, however its better to be in appended string form here, so that in future if there is already set value, it should carry that in final string.

File: scripts/minikube.sh [hunk 8]
```diff
@@ -202,6 +211,14 @@ up)
 
     disable_storage_addons
 
+    #  get kubernetes version we are operating on and accordingly enable feature gates
+    KUBE_MAJOR=$(kube_version 1)
+    KUBE_MINOR=$(kube_version 2)
+    if  [ "${KUBE_MAJOR}" -eq 1 ] && [ "${KUBE_MINOR}" -ge 20 ]
+    then
+    # if kubernetes version is greater than 1.20 enable Generic Ephemeral Volume feature gate
+    K8S_FEATURE_GATES="${K8S_FEATURE_GATES},GenericEphemeralVolume=true"
```
Comment (nixpanic) [inline scripts/minikube.sh]: if `K8S_FEATURE_GATES` is an empty string, maybe the extra `,` can cause issues?

File: scripts/minikube.sh [hunk 9]
```diff
@@ -160,7 +169,7 @@ CSI_RESIZER_VERSION=${CSI_RESIZER_VERSION:-"v1.2.0"}
 CSI_NODE_DRIVER_REGISTRAR_VERSION=${CSI_NODE_DRIVER_REGISTRAR_VERSION:-"v2.2.0"}
 
 #feature-gates for kube
-K8S_FEATURE_GATES=${K8S_FEATURE_GATES:-"ExpandCSIVolumes=true"}
+K8S_FEATURE_GATES=${K8S_FEATURE_GATES:-""}
```
Comment (nixpanic) [inline scripts/minikube.sh:163]: Now you are dropping `ExpandCSIVolumes` without comment. It is better to do that in a stand-alone commit. As a cleanup, it can be included in this PR, no need to create a separate one.

**Review by nixpanic — DISMISSED** [review]

File: scripts/minikube.sh (hunk 9 above)
Comment (humblec) [reply scripts/minikube.sh:163]: sure, put it as a seperate commit here in this PR itself.

File: scripts/minikube.sh (hunk 8 above)
Comment (humblec) [reply scripts/minikube.sh]: as discussed, its not causing issue rather its going through, so will keep this version.

**Comment by humblec** [comment]
@Mergifyio rebase

**Comment by humblec** [comment]
@Mergifyio refresh

**Comment by humblec** [comment]
Test failure https://github.com/ceph/ceph-csi/issues/2610

**Comment by humblec** [comment]
@Madhu-1 can you please review this ?

File: e2e/cephfs.go [hunk 10]
```diff
@@ -312,6 +313,36 @@ var _ = Describe("cephfs", func() {
...
 				})
 			}
+			By("verify generic ephemeral volume support", func() {
+				// generic ephemeral volume support is beta since v1.21.
+				if !k8sVersionGreaterEquals(f.ClientSet, 1, 21) {
+					Skip("generic ephemeral volume only supported from v1.21+")
+				}
+				err := createCephfsStorageClass(f.ClientSet, f, true, nil)
+				if err != nil {
+					e2elog.Failf("failed to create CephFS storageclass with error %v", err)
```
Comment (Madhu-1) [inline e2e/cephfs.go]: `failed to create CephFS storageclass with error %v` to `failed to create CephFS storageclass: %v` `with error` formatting needs to be adjusted in other places also

File: e2e/cephfs.go [hunk 11]
```diff
@@ -312,6 +313,36 @@ var _ = Describe("cephfs", func() {
...
+				// create application
+				app, err := loadApp(appEphemeralPath)
+				if err != nil {
+					e2elog.Failf("failed to load application: %v", err)
+				}
+				app.Namespace = f.UniqueName
+				err = createApp(f.ClientSet, app, deployTimeout)
+				if err != nil {
+					e2elog.Failf("failed to create application: %v", err)
+				}
```
Comment (Madhu-1) [inline e2e/cephfs.go:334]: add a subvolumeCount validation here to make sure subvolume is present in backend?

File: e2e/rbd.go [hunk 12]
```diff
@@ -367,6 +368,40 @@ var _ = Describe("RBD", func() {
...
+				err = waitToRemoveImagesFromTrash(f, defaultRBDPool, deployTimeout)
+				if err != nil {
+					e2elog.Failf("failed to validate rbd images in pool %s trash: %v", defaultRBDPool, err)
+				}
+				err = waitToRemoveImagesFromTrash(f, defaultRBDPool, deployTimeout)
+				if err != nil {
+					e2elog.Failf("failed to validate rbd images in pool %s trash: %v", defaultRBDPool, err)
+				}
+			})
+
```
Comment (Madhu-1) [inline e2e/rbd.go]:
```suggestion
// validate images in trash
				err = waitToRemoveImagesFromTrash(f, defaultRBDPool, deployTimeout)
				if err != nil {
					e2elog.Failf("failed to validate rbd images in pool %s trash: %v", defaultRBDPool, err)
				}
```

File: scripts/minikube.sh [hunk 13]
```diff
@@ -202,6 +211,14 @@ up)
 
     disable_storage_addons
 
+    #  get kubernetes version we are operating on and accordingly enable feature gates
+    KUBE_MAJOR=$(kube_version 1)
+    KUBE_MINOR=$(kube_version 2)
+    if  [ "${KUBE_MAJOR}" -eq 1 ] && [ "${KUBE_MINOR}" -ge 21 ]
+    then
+    # if kubernetes version is greater than 1.21 enable Generic Ephemeral Volume feature gate
+    K8S_FEATURE_GATES="${K8S_FEATURE_GATES},GenericEphemeralVolume=true"
```
Comment (Madhu-1) [inline scripts/minikube.sh]: as it is beta in 1.21+ already enabled by default. we don't need to enable it explicitly?

File: e2e/cephfs.go [hunk 14]
```diff
@@ -312,6 +313,36 @@ var _ = Describe("cephfs", func() {
 					}
 				})
 			}
+			By("verify generic ephemeral volume support", func() {
+				// generic ephemeral volume support is beta since v1.21.
+				if !k8sVersionGreaterEquals(f.ClientSet, 1, 21) {
```
Comment (Madhu-1) [inline e2e/cephfs.go:318]: it's skipping for 1.20 but as per https://github.com/ceph/ceph-csi/pull/2603#discussion_r739026205 it's agreed to run on  1.20 also?

Comment (humblec) [reply e2e/cephfs.go:318]: this was deliberate, considering this is beta and supported since 1.21, this should be good I believe.

File: scripts/minikube.sh (hunk 13 above)
Comment (humblec) [reply scripts/minikube.sh]: @Madhu-1 if this script is executed for versions less than 1.21, we dont need to enable this feature gate, thats why tihs check in place.

File: e2e/rbd.go (hunk 5 above)
Comment (humblec) [reply e2e/rbd.go:377]: on more thoughts, considering other tests are enabled for beta for example, snapshot/clone...etc it looks like its fine to enable only on versions 1.21. also next release is anyway going to throw away 1.20 from here, so should be good I believe. Last patch refresh adjusted the same.

**Comment by humblec** [comment]
@nixpanic @Madhu-1 all the comments are addressed, ptal.. thanks in advance.

File: e2e/cephfs.go (hunk 14 above)
Comment (nixpanic) [reply e2e/cephfs.go:318]: agreed, but in that case it is not needed to enable the featue gate in `minikube.sh`?

Comment (humblec) [reply e2e/cephfs.go:318]: it was added considering the script can be used independently for kube deployments, but may be here, we can skip that and only have e2e check. adjusted the PR..

File: scripts/minikube.sh (hunk 13 above)
Comment (humblec) [reply scripts/minikube.sh]: it was added considering the script can be used independently for kube deployments, but may be here, we can skip that and only have e2e check. adjusted the PR..

File: examples/rbd/pod-ephemeral.yaml [hunk 15]
```diff
@@ -0,0 +1,22 @@
...
+      image: docker.io/library/nginx:latest
+      volumeMounts:
+        - mountPath: /myspace
+          name: mypvc
+  volumes:
+    - name: mypvc
+      ephemeral:
+        volumeClaimTemplate:
+          spec:
+            accessModes: [ReadWriteOnce]
```
Comment (nixpanic) [inline examples/rbd/pod-ephemeral.yaml]: The other examples use the cleaner (IMHO) format of

```yaml
accessModes:
  - ReadWriteOnce
```

it is probably better to use the same formatting in all examples we have

**Review by nixpanic — DISMISSED** [review]
just a nit about the yaml formatting of the examples, looks good to me otherwise

File: examples/rbd/pod-ephemeral.yaml [hunk 16]
```diff
@@ -0,0 +1,23 @@
+--- 
+apiVersion: v1
+kind: Pod
+metadata:
+  name: csi-rbd-demo-ephemeral-pod
+spec:
+  containers:
+      image: "docker.io/library/nginx:latest"
```
Comment (Madhu-1) [inline examples/rbd/pod-ephemeral.yaml]: remove `""` from image name

File: examples/rbd/pod-ephemeral.yaml [hunk 17]
```diff
@@ -0,0 +1,23 @@
...
+      ephemeral:
+        volumeClaimTemplate:
+          spec:
+            accessModes:
+              - ReadWriteOnce
+            resources:
+              requests:
+                storage: 1Gi
+            storageClassName: csi-rbd-sc
+      name: mypvc
```
Comment (Madhu-1) [inline examples/rbd/pod-ephemeral.yaml]: move the name to line 14  and keep it similar to cephfs

**Comment by humblec** [comment]
@Mergifyio rebase

**Comment by humblec** [comment]
@nixpanic @Madhu-1 ptal.. thanks.

**Review by Madhu-1 — DISMISSED** [review]

**Comment by humblec** [comment]
rebased to devel, and also a minor change ( as pointed out by @Rakshith-R in ci logs)  from `appEphemeralPath = cephFSExamplePath + "pod-ephemeral.yaml"` to `appEphemeralPath := cephFSExamplePath + "pod-ephemeral.yaml"`

**Review by Madhu-1 — APPROVED** [review]

**Review by Rakshith-R — APPROVED** [review]

**Comment by humblec** [comment]
@Mergifyio rebase

**Comment by nixpanic** [comment]
> /retest ci/centos/upgrade-tests

For some reason the CentOS jobs were not listed, this causes a refresh from Jenkins in the CentOS CI.

**Comment by humblec** [comment]
@Mergifyio rebase

**Comment by humblec** [comment]
@Mergifyio rebase

**Comment by humblec** [comment]
@Mergifyio refresh

**Comment by humblec** [comment]
@Mergifyio rebase

**Comment by Rakshith-R** [comment]
@mergifyio rebase

**Comment by humblec** [comment]
@Mergifyio rebase

**Comment by humblec** [comment]
@Mergifyio rebase
